### PR TITLE
API extension points

### PR DESCRIPTION
**Pull request by treeder**
This enables the user to add new endpoints to the API, that aren't part of the default API. Also simplifies main down to just a few lines as the default. 

See this for usage: https://github.com/iron-io/functions/compare/master...extensions2#diff-227391467e160aca7abf1b0671a0f807

Part of #462.

### CODE REVIEW

**Comment by ccirello** [comment]
OK. `AppListener`. From my perspective, this is a long story. It starts as a single interface for all Listeners types, and then iteratable through all Listeners. Sometime ago, I believe in https://github.com/iron-io/functions/pull/412 , @pedronasser changed it - by my suggestion - to break into smaller interfaces, and only iterate each Listener type. This PR is reverting it.

@pedronasser any insights on this change?

**Comment by treeder** [comment]
I can see the benefit of the smaller interfaces from a user perspective, but also felt more complex, more things to know about. And this is only for apps, we'll need the same thing for routes and whatever else a user will be able to do in the future.

File: api/const.go [hunk 1]
```diff
@@ -4,4 +4,8 @@ package api
 const (
 	AppName string = "app_name"
```
Comment (pedronasser) [inline api/const.go]: Why these doesn't have C prefix?

Comment (ccirello) [inline api/const.go]: Why add a C prefix first off? This is not a go idiomatic standard.

Comment (treeder) [inline api/const.go]: @ccirello removed the C and I didn't feel like adding them back. ;)  Can figure out how we want to name the constants later.

**Comment by treeder** [comment]
@ccirello @seiflotfy @pedronasser could someone review this please.

File: api/server/server.go [hunk 2]
```diff
@@ -53,6 +54,24 @@ type Server struct {
 
 const cacheSize = 1024
 
+// NewEnv creates a new IronFunctions server based on env vars.
+func NewEnv(ctx context.Context) *Server {
```
Comment (pedronasser) [inline api/server/server.go]: Wouldn't be more descriptive to use the name `NewFromEnv`?

File: docs/operating/extending.md [hunk 3]
```diff
@@ -15,7 +15,7 @@ package main
 
 import (
     "context"
-
+df
```
Comment (pedronasser) [inline docs/operating/extending.md]: Should this be removed?

File: api/server/special_handler_test.go [hunk 4]
```diff
@@ -1,64 +1,51 @@
...
 
 type testSpecialHandler struct{}
 
 func (h *testSpecialHandler) Handle(c HandlerContext) error {
-	c.Set(api.AppName, "test")
+	// c.Set(api.AppName, "test")
 	return nil
 }
 
 func TestSpecialHandlerSet(t *testing.T) {
```
Comment (ccirello) [inline api/server/special_handler_test.go]: perhaps adding t.Skip()?

File: api/server/server.go (hunk 2 above)
Comment (treeder) [inline api/server/server.go]: Ya, good call, changing.

**Comment by seiflotfy** [comment]
Really nice work! LGTM!

**Pull request merged**